tour_model_eval/build_save_model.py

```diff
@@ -0,0 +1,170 @@
+                # collect location features of the bin from the first round of clustering
+                # feat[0:4] are start/end coordinates
+                bin_loc_feat[str(fl)] = [feat[0:4] for feat in x]
+                # here we pass in features(x) from selected second round trips to build the model
+                method = 'single'
+                clusters = lp.get_second_labels(x, method, low, dist_pct)
+                n_clusters = len(set(clusters))
+                # build the model
+                kmeans = sc.KMeans(n_clusters=n_clusters, random_state=0).fit(x)
+                # collect all models, the key is the label from the 1st found
+                # e.g.{'0': KMeans(n_clusters=2, random_state=0)}
+                model_coll[str(fl)] = kmeans
+                # get labels from the 2nd round of clustering
+                second_labels = kmeans.labels_
+
+                first_label_obj = []
+
+                # save user labels for every cluster
+                second_label_set = list(set(second_labels))
+                sec_round_labels = {}
+                for sl in second_label_set:
+                    sec_sel_trips = []
+                    sec_label_obj = []
+                    for idx, second_label in enumerate(second_labels):
+                        if second_label == sl:
+                            sec_sel_trips.append(second_round_trips[idx])
+                    user_label_df = pd.DataFrame([trip['data']['user_input'] for trip in sec_sel_trips])
+                    user_label_df = lp.map_labels(user_label_df)
+                    # compute the sum of trips in this cluster
+                    sum_trips = len(user_label_df)
+                    # compute unique label sets and their probabilities in one cluster
+                    # 'p' refers to probability
+                    unique_labels = user_label_df.groupby(user_label_df.columns.tolist()).size().reset_index(name='count')
+                    unique_labels['p'] = [unique_labels.iloc[i]['count'] / sum_trips for i in range(len(unique_labels))]
+                    labels_columns = user_label_df.columns.to_list()
+                    one_set_labels = {}
+                    for i in range(len(unique_labels)):
+                        # e.g. labels_only={'mode_confirm': 'pilot_ebike', 'purpose_confirm': 'work', 'replaced_mode': 'walk'}
+                        labels_only = {column: unique_labels.iloc[i][column] for column in labels_columns}
+                        one_set_labels["labels"] = labels_only
+                        one_set_labels['p'] = unique_labels.iloc[i]['p']
+                        # e.g. one_set_labels = {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 1.0}
+                        # in case append() method changes the dict, we use deepcopy here
+                        labels_set = copy.deepcopy(one_set_labels)
+                        sec_label_obj.append(labels_set)
```

Comment:
line 126 `append()`
This will change the dict if the dict is not an entire new object. if I directly append `one_set_labels`, the values in `one_set_labels` will be changed.

Reply:
really? because that is not how python dicts or lists work. see below.
If `one_set_labels` is changed, there is some other reason, you need to understand and explain it.

```
>>> sec_label_obj = []
>>> one_set_labels = {}
>>> one_set_labels["labels"] = {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}
>>> one_set_labels["p"] = 1.0
>>> sec_label_obj.append(one_set_labels)
>>> sec_label_obj
[{'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 1.0}]
>>> one_set_labels
{'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 1.0}
```

Reply:
Here you are using one dict. But as you append more dicts with same name(`one_set_labels`), the previous added dict will be changed.
```
>>> test_list = []
>>> a={'q':1,'w':[]}
>>> test_list.append(a)
>>> print(test_list)
[{'q': 1, 'w': []}]

>>> a['q']=2
>>> a['w'].append(123)
>>> test_list.append(a)
>>> print(test_list)
[{'q': 2, 'w': [123]}, {'q': 2, 'w': [123]}]
```

Several days ago, I found some explanations in Chinese. Someone says that what appended to the list is the address of the dict. It doesn't create a new object in the list. So, every time we append a dict with the same name, it actually points to the same address. 

Also, there is a difference between `copy()` and `deepcopy()`
`copy()` is shallow copy
```
>>> a={'q':1,'w':[]}
>>> b=a.copy()
>>> b['q']=2
>>> b['w'].append(123)
>>> print(a)
>>> print(b)
{'q': 1, 'w': [123]}
{'q': 2, 'w': [123]}
```
If we change to `deepcopy()`, we can have entirely new object.
```
>>> a={'q':1,'w':[]}
>>> b=copy.deepcopy(a)
>>> b['q']=2
>>> b['w'].append(123)
>>> print(a)
>>> print(b)
{'q': 1, 'w': []}
{'q': 2, 'w': [123]}

Reply:
Here you are using one dict. But as you append more dicts with same name(one_set_labels), the previous added dict will be changed. This is not true if you do it "pythonically"

```
>>> sec_label_obj = []
>>> for p in range(0,10,1):
...     one_set_labels = {}
...     one_set_labels["labels"] = {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}
...     one_set_labels["p"] = p/10
...     sec_label_obj.append(one_set_labels)
...
>>> sec_label_obj
[{'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.0}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.1}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.2}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.3}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.4}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.5}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.6}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.7}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.8}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.9}]
```

This style is more pythonic because it lends itself to list comprehensions, use with itertools, functional style programming, etc

```
>>> sec_label_obj = [{"labels": {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, "p": p/10} for p in range(0,10,1)]
>>> sec_label_obj
[{'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.0}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.1}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.2}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.3}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.4}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.5}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.6}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.7}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.8}, {'labels': {'mode_confirm': 'walk', 'replaced_mode': 'walk', 'purpose_confirm': 'exercise'}, 'p': 0.9}]
```

> Several days ago, I found some explanations in Chinese. Someone says that what appended to the list is the address of the dict. It doesn't create a new object in the list. So, every time we append a dict with the same name, it actually points to the same address.

@corinne-hcr I am aware of the difference between shallow copy and deepcopy. But the difference is primarily when you **modify** objects that you have appended to the list (e.g. `b['q']=2`); merely appending the object doesn't change it. In general modifying objects that you are appending to the list is bad practice because the code cannot be easily refactored using the alternate styles I outlined above.

Where are you modifying objects in your code?

Reply:
I think I understand the differences between your output and mine. I created `one_set_labels = {}` before the loop, then I didn't have a new object every time I change the values. You created `one_set_labels = {}` in the loop, so you have a new object every time. But I am going to change the structure. Glad to figure it out!

Reply:
Creating one object outside the loop and then deepcopying in the loop is not pythonic. It will make it harder to maintain the code in the future.

Reply:
@corinne-hcr I still see the `deepcopy` here